### PR TITLE
feat: Adds user detection for multiple user agents and fixes tool reduplication

### LINKED ISSUE #1664: [Feature]: Add integration user agent detection and tool deduplication for Gemini CLI, Qwen CLI, Cursor, Codex, and n8n
### Prerequisites

- [x] I have searched existing issues and discussions to avoid duplicates

### Problem to solve

Currently, Bifrost only has tool deduplication logic for Claude Code (Claude CLI). This causes tools to be attached twice when Bifrost is used as both an LLM and MCP gateway with other CLI tools/integrations.

## Current State

1. Detecting the `claude-cli` user agent from headers in the Anthropic integration
2. Setting `BifrostContextKeyUserAgent` to track it
3. Building a duplicate check map in `buildIntegrationDuplicateCheckMap()` to handle the `mcp__{foreign_name}__{tool_name}` pattern that Claude CLI uses

### Proposed solution

Extend the deduplication mechanism to support:
   - **Gemini CLI** - detect `gemini-cli` user agent
   - **Qwen CLI** - detect `qwen-cli` user agent
   - **Cursor** - detect `cursor` user agent
   - **Codex** - detect `codex` user agent
   - **n8n** - detect `n8n` user agent

   ## Implementation Tasks

   1. **Update integration detection** (e.g., in anthropic integration and other providers):
      - Check User-Agent headers for each CLI tool
      - Set `BifrostContextKeyUserAgent` with the appropriate value
      - May need to handle these in multiple provider integrations (Anthropic, Google, OpenAI, etc.)

   2. **Extend deduplication logic** in `core/mcp/toolmanager.go`:
      - Add cases to `buildIntegrationDuplicateCheckMap()` for each new integration
      - Document the naming pattern each CLI uses (similar to Claude CLI's `mcp__{foreign_name}__{tool_name}`)
      - Handle extraction of tool names from existing tools for duplicate checking

   3. **Testing**:
      - Test with each CLI tool to verify no duplicate tools are attached
      - Verify that tools are properly discovered when Bifrost is used as both LLM and MCP gateway

   ## Related Code

   - `core/mcp/toolmanager.go`: Lines 233-279 (`buildIntegrationDuplicateCheckMap`)
   - `core/mcp/toolmanager.go`: Lines 301-309 (context key lookup)
   - `transports/bifrost-http/integrations/anthropic.go`: Lines 290-293 (Claude CLI detection example)

   ## Notes

   Each CLI tool may use different naming patterns for tools, so we need to research or document:
   - What pattern each tool uses for MCP tool naming
   - How to extract the original tool name from the prefixed version
   - Whether they all follow the same `prefix__{name}` pattern or use different formats

### Alternatives considered

_No response_

### Area(s)

Core (Go)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by roroghost17**
## Summary

Enhances MCP tool deduplication to support multiple client integrations (Claude CLI, Gemini CLI, Codex CLI, Qwen CLI, OpenCode) by implementing integration-aware duplicate detection that handles different tool naming patterns used by each client.

Fixes #1664

## Changes

- Added `UserAgentIdentifiers` type and constants for different CLI clients in `core/schemas/useragents.go`
- Enhanced `buildIntegrationDuplicateCheckMap` to handle tool naming patterns for each integration:
    - Claude CLI: `mcp__{server}__{tool_name}` (double underscores)
    - Gemini CLI: `mcp_{server}_{tool_name}` (single underscore)
    - Codex CLI: `mcp__{server}__{tool_name}` with hyphens converted to underscores
    - Qwen CLI: `mcp__{server}__{tool_name}` (double underscores, hyphens preserved)
    - OpenCode: `{server}_{tool_name}` (no mcp prefix, hyphens preserved)
    - Cursor: currently does not need dedupllication
    - n8n: Currently does not need, but could be needed if https://github.com/n8n-io/n8n/issues/25269 this gets resolved
- Added `integrationDuplicateCheck` function to handle Codex CLI's hyphen-to-underscore normalization
- Implemented `ExtractAndSetUserAgentFromHeaders` to automatically detect and set user agent context
- Updated integration endpoints to use the new user agent extraction
- Added comprehensive test suite covering all integration patterns and edge cases

## Type of change

- [x] Feature
- [x] Refactor

## Affected areas

- [x] Core (Go)
- [x] Transports (HTTP)
- [x] Providers/Integrations

## How to test

Run the new test suite to verify deduplication logic:

```sh
go test ./core/mcp -v -run TestBuildIntegrationDuplicateCheckMap
go test ./core/mcp -v -run TestParseAndAddToolsToRequest
```

Test with different client user agents by setting appropriate headers when making requests to verify tools are not duplicated.

## Screenshots/Recordings

N/A

## Breaking changes

- [ ] Yes
- [x] No

## Related issues

N/A

## Security considerations

No security implications - this change only affects tool deduplication logic and user agent detection.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [ ] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [x] I verified the CI pipeline passes locally if applicable

### CODE REVIEW

**Comment by roroghost17** [comment]
* **#2849** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2849?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2796** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2796?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2795** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2795?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/2795?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: core/mcp/toolmanager_test.go [hunk 1]
```diff
@@ -0,0 +1,935 @@
...
+	}
+
+	m := buildIntegrationDuplicateCheckMap(tools, codexUA, defaultLogger)
+
+	for _, want := range []string{"bash", "list_files"} {
+		if !m[want] {
+			t.Errorf("Codex: expected %q in duplicate map", want)
+		}
+	}
+}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply core/mcp/toolmanager_test.go:430]: the names have been cross-checked and added based on what the agents actually send.

File: core/mcp/toolmanager.go [hunk 2]
```diff
@@ -296,14 +296,112 @@ func buildIntegrationDuplicateCheckMap(existingTools []schemas.ChatTool, integra
...
+				if underscoreIdx != -1 && underscoreIdx < len(existingToolName)-1 {
+					toolName := existingToolName[underscoreIdx+1:]
+					if toolName != "" {
+						duplicateCheckMap[toolName] = true
+						duplicateCheckMap[existingToolName] = true
+					}
+				}
+			}
+		}
 	}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply core/mcp/toolmanager.go:388]: this is checked based on all the patterns that have been seen with individual tests on all agents

File: core/mcp/toolmanager.go [hunk 3]
```diff
@@ -296,14 +296,112 @@ func buildIntegrationDuplicateCheckMap(existingTools []schemas.ChatTool, integra
...
+				underscoreIdx := strings.Index(existingToolName, "_")
+				if underscoreIdx != -1 && underscoreIdx < len(existingToolName)-1 {
+					toolName := existingToolName[underscoreIdx+1:]
+					if toolName != "" {
+						duplicateCheckMap[toolName] = true
+						duplicateCheckMap[existingToolName] = true
+					}
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply core/mcp/toolmanager.go:387]: the patterns are based on what all user agents actually send

**Review by akshaydeo — DISMISSED** [review]

File: core/schemas/useragents.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import "strings"
+
+// UserAgentIdentifiers lists substrings that may appear in User-Agent for a given integration.
+// Versions of the same client may use different strings; Matches checks any of them.
+type UserAgentIdentifiers []string
+
+var (
+	// ClaudeCLI — Anthropic Claude Code / Claude CLI (identifiers vary by release).
+	ClaudeCLI   = UserAgentIdentifiers{"claude-cli", "claude-code"}
+	GeminiCLI   = UserAgentIdentifiers{"GeminiCLI"}
```
Comment (akshaydeo) [inline core/schemas/useragents.go]: can we just keep lowercase checks across differet variants?

Comment (roroghost17) [reply core/schemas/useragents.go]: updated.

File: core/schemas/useragents.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	CodexCLI    = UserAgentIdentifiers{"codex-tui"}
+	QwenCodeCLI = UserAgentIdentifiers{"QwenCode"}
+	OpenCode    = UserAgentIdentifiers{"opencode"}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply core/schemas/useragents.go]: we only handle the actual agent strings that the CLIs send. This has been checked manually and added.

File: core/schemas/useragents.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+			break
+		}
+	}
+	if len(userAgent) > 0 {
+		ua := userAgent[0]
+		for _, ids := range integrationUserAgents {
+			if ids.Matches(ua) {
+				bifrostCtx.SetValue(BifrostContextKeyUserAgent, ua)
+				break
+			}
```
[the remark this answers is not in the document]
Comment (roroghost17) [reply core/schemas/useragents.go:69]: there can't be multiple user agents

**Review by akshaydeo — APPROVED** [review]

**Comment by akshaydeo** [comment]
### Merge activity

* **Apr 20, 2:07 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/2795).
* **Apr 20, 2:08 PM UTC**: @akshaydeo merged this pull request with [Graphite](https://app.graphite.com/github/pr/maximhq/bifrost/2795).
